cvmfs/uid_map.h

```diff
@@ -0,0 +1,183 @@
+  typedef T                                       value_type;
+  typedef typename std::map<key_type, value_type> map_type;
+
+ public:
+  IntegerMap()
+    : valid_(true)
+    , has_default_value_(false)
+    , default_value_(T(0)) {}
+
+  /**
+   * Define a mapping from k to v
+   * @param k  map the given value to v
+   * @param v  the value given in k is mapped to this
+   **/
+  void Set(const T k, const T v) { map_[k] = v; }
+
+  /**
+   * Sets a default (or fallback) value to be used if no other mapping rule fits
+   * Note: A previously defined default value is overwritten.
+   * @param v  the value to be used as a fallback in Map()
+   */
+  void SetDefault(const T v) {
+    has_default_value_ = true;
+    default_value_     = v;
+  }
+
+  /**
+   * Reads mapping rules from a provided file path. The file format is discussed
+   * in the class description above.
+   * Note: If a read failure occurs the IntegerMap<> is declared invalid and
+   *       must not be used anymore.
+   *
+   * @param path  the file path to be read
+   * @return      true if the file was successfully read
+   */
+  bool Read(const std::string &path) {
+    valid_ = ReadFromFile(path);
+    return IsValid();
+  }
+
+  /**
+   * Checks if a mapping rule for a given value is available
+   * @param k  the value to be checked
+   * @return   true if a mapping rule for k exists
+   */
+  bool Contains(const T k) const {
+    assert(IsValid());
+    return map_.find(k) != map_.end();
+  }
+
+  /**
+   * Applies the mapping rules inside this IntegerMap<> to the given value.
+   * @param k  the value to be mapped
+   * @return   the result of the mapping rule application (might be the default)
+   */
+  T Map(const T k) const {
+    assert(IsValid());
+    typename map_type::const_iterator i = map_.find(k);
+    if (i != map_.end()) {
+      return i->second;
+    }
+
+    return (HasDefault())
+      ? default_value_
+      : T(0);
+  }
+
+  bool   IsValid()    const { return valid_;             }
+  bool   HasDefault() const { return has_default_value_; }
+  size_t RuleCount()  const { return map_.size();        }
+
+  T GetDefault() const { assert(has_default_value_); return default_value_; }
+  const map_type& GetRuleMap() const { return map_; }
+
+ protected:
+  bool ReadFromFile(const std::string &path) {
+    FILE *fmap = fopen(path.c_str(), "r");
+    if (!fmap) {
+      LogCvmfs(kLogUtility, kLogDebug, "failed to open %s (errno: %d)",
+               path.c_str(), errno);
+      return false;
+    }
+
+    std::string line;
+    unsigned int line_number = 0;
+    while (GetLineFile(fmap, &line)) {
+      ++line_number;
+      line = Trim(line);
+      if (line.empty() || line[0] == '#') {
+        continue;
+      }
+
+      std::vector<std::string> components = SplitString(line, ' ');
+      FilterEmptyStrings(&components);
+      if (components.size() != 2    ||
+          !IsNumeric(components[1]) ||
```

Comment:
Fixed by using `sanitizer::IntegerSanitizer`. That will allow negative UIDs which are apparently common on some UNIX systems like HP-UX. Questionable if we will ever support that, but who knows.